shared/src/main/scala/coop/rchain/pcasper/finalization/lazy/LazyFinalizer.scala

```diff
@@ -0,0 +1,95 @@
+package coop.rchain.pcasper.finalization.`lazy`
+
+import cats.effect.Sync
+import cats.syntax.all._
+import coop.rchain.pcasper.finalization._
+
+final case class LazyFinalizer[F[_]: Sync, M, S, P](
+    sndr: S,
+    view: Map[S, M],
+    bonds: Map[S, Long],
+    dag: DagData[F, M, S]
+) extends Finalizer[F, M, S, LazyFinal[M, S]] {
+  import dag._
+
+  val viewBoundary = view.mapValues(seqNum)
+  val witnessesInViewF =
+    witnessesF(_: M)
+      .map(_.filter {
+        case (s, m) =>
+          val latestInView = viewBoundary.get(s)
+          assert(
+            latestInView.isDefined,
+            s"Sender $s is not present in latest messages defining the view $view."
+          )
+          seqNum(m) <= latestInView.get
+      }.toMap)
+  val dagInView = new DagData[F, M, S] {
+    override def witnessesF: M => F[Map[S, M]]      = witnessesInViewF
+    override def justificationsF: M => F[Map[S, M]] = dag.justificationsF
+    override def seqNum: M => Long                  = dag.seqNum
+    override def sender: M => S                     = dag.sender
+  }
+
+  private def nextFringeOpt(
+      base: Fringe[M, S],
+      constraint: Set[S] => Boolean = _ => true
+  ): F[Option[Fringe[M, S]]] = {
+
+    val findAdvancement = LazyPartitioner(dagInView, bonds, constraint).findPartition(base)
+
+    // advance fringe if there are justifications higher
+    def jumpToJss(fringe: Map[S, M]): F[Map[S, M]] =
+      for {
+        // across justifications of fringe advancement
+        advJss <- fringe.valuesIterator.toVector.traverse(justificationsF)
+        // find in justification of advanced senders the higher messages outside advanced senders
```

Comment:
What is `advanced senders`?

Reply:
I can't say now because the algorithm and terminology are not clear for me at the moment. You can leave the comment as is.